locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,192 @@
+            "The length of time to wait for a Softlayer VLAN ID", Duration.minutes(10));
+
+    /** Caches the semaphore object for locking access to the scope. */
+    private static final Cache<String, Semaphore> semaphoreCache = CacheBuilder.newBuilder().build();
+
+    /** Caches the public VLAN id that should be shared between all entities in the same scope. */
+    private static final Cache<String, Integer> publicVlanIdCache = CacheBuilder.newBuilder().build();
+
+    /** Caches the private VLAN id that should be shared between all entities in the same scope. */
+    private static final Cache<String, Integer> privateVlanIdCache = CacheBuilder.newBuilder().build();
+
+    /**
+     * Used to obtain the VLANs being used by the first created {@link JcloudsMachineLocation}.
+     */
+    @Override
+    public void customize(JcloudsLocation location, ComputeService computeService, JcloudsMachineLocation machine) {
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME))) {
+            String message = String.format("Invalid location provider: %s", provider);
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        Semaphore semaphore = null;
+        String scopeUid = getScopeUid(location);
+
+        synchronized (SoftLayerSameVlanLocationCustomizer.class) {
+            semaphore = semaphoreCache.getIfPresent(scopeUid);
+            if (semaphore == null) {
+                throw new IllegalStateException("No semaphore available for scope: " + scopeUid);
+            }
+        }
+
+        Integer publicVlanId = publicVlanIdCache.getIfPresent(scopeUid);
+        Integer privateVlanId = privateVlanIdCache.getIfPresent(scopeUid);
+        if (privateVlanId != null || publicVlanId != null) {
+            LOG.debug("SoftLayer VLANs already configured for application: {}", scopeUid);
+            return;
+        }
+
+        // Ask SoftLayer API for the VLAN details for this VM
+        VirtualGuestApi api = computeService.getContext().unwrapApi(SoftLayerApi.class).getVirtualGuestApi();
+        Long serverId = Long.parseLong(machine.getJcloudsId());
+        VirtualGuest guest = api.getVirtualGuestFiltered(serverId,
+                "primaryNetworkComponent;" +
+                "primaryNetworkComponent.networkVlan;" +
+                "primaryBackendNetworkComponent;" +
+                "primaryBackendNetworkComponent.networkVlan");
+        publicVlanId = guest.getPrimaryNetworkComponent().getNetworkVlan().getVlanNumber();
+        privateVlanId = guest.getPrimaryBackendNetworkComponent().getNetworkVlan().getVlanNumber();
+
+        // Save the VLAN ids and release the semaphore
+        publicVlanIdCache.put(scopeUid, publicVlanId);
+        privateVlanIdCache.put(scopeUid, privateVlanId);
+        semaphore.release();
```

Comment:
if using `semaphore with owners` we could check whether we are the thread which acquired the semaphore and log a warning if not, for good measure.

i think `release` whether or not we were the creator thread, because we now have a VM others can use.

however if the semaphore already has a permit available then don't release (eg if we're rebinding).